lib/IRGen/GenCall.cpp

```diff
@@ -1685,18 +1688,21 @@ static void emitCoerceAndExpand(IRGenFunction &IGF,
 }
 
 static void emitDirectExternalArgument(IRGenFunction &IGF, SILType argType,
-                                       llvm::Type *toTy, Explosion &in,
-                                       Explosion &out) {
+                                       const clang::CodeGen::ABIArgInfo &AI,
+                                       Explosion &in, Explosion &out) {
   // If we're supposed to pass directly as a struct type, that
   // really means expanding out as multiple arguments.
-  ArrayRef<llvm::Type *> expandedTys = expandScalarOrStructTypeToArray(toTy);
+  llvm::Type *coercedTy = AI.getCoerceToType();
+  ArrayRef<llvm::Type *> expandedTys =
+      expandScalarOrStructTypeToArray(coercedTy);
 
   auto &argTI = cast<LoadableTypeInfo>(IGF.getTypeInfo(argType));
   auto inputSchema = argTI.getSchema();
 
   // Check to see if we can pairwise coerce Swift's exploded scalars
   // to Clang's expanded elements.
-  if (canCoerceToSchema(IGF.IGM, expandedTys, inputSchema)) {
```

Comment:
Sorry for noticing this so late but this is not NFC for darwin.

We can see this in the tests. They should not change (but there are added extra loads/stores of i8*)

I believe what is going wrong is that this check should be something like (indirect arguments should be coercible):
if (
 (AI.direct() && AI.getCanBeFlattened()) || 
  !AI.direct()) &&
canCoerceToSchema(...))

Reply:
Hmm, I thought that @rjmccall was okay with the additional coercion (since it should get optimized out).  In any case, I'll see if that works for avoiding the unnecessary memory traffic.

Reply:
Well, if we can avoid coercing through memory, great.

Reply:
Adjusted to avoid the coercion through memory unless needed.  I expect that the linux build should be fine (it passed locally).